geode-wan/src/main/java/org/apache/geode/cache/client/internal/locator/wan/LocatorMembershipListenerImpl.java

```diff
@@ -213,4 +199,119 @@ public void clearLocatorInfo() {
+                  targetLocator.getHostName(), targetLocator.getPort(), 1, exception});
+        }
+
+        if (!failedMessages.containsKey(targetLocator)) {
+          failedMessages.put(targetLocator, new HashSet<>());
+        }
+
+        failedMessages.get(targetLocator).add(locatorJoinMessage);
+      }
+    }
+
+    boolean retryMessage(DistributionLocatorId targetLocator, LocatorJoinMessage locatorJoinMessage,
+        int retryAttempt) {
+      DistributionLocatorId advertisedLocator = locatorJoinMessage.getLocator();
+
+      try {
+        tcpClient.requestToServer(targetLocator.getHost(), locatorJoinMessage, memberTimeout,
+            false);
+
+        return true;
+      } catch (Exception exception) {
+        if (retryAttempt == WAN_LOCATOR_DISTRIBUTION_RETRY_ATTEMPTS) {
+          logger.warn(LISTENER_FINAL_FAILURE_MESSAGE,
+              new Object[] {advertisedLocator.getHostName(), advertisedLocator.getPort(),
+                  targetLocator.getHostName(), targetLocator.getPort(), retryAttempt, exception});
+        } else {
+          if (logger.isDebugEnabled()) {
+            logger.debug(LISTENER_FAILURE_MESSAGE,
+                new Object[] {advertisedLocator.getHostName(), advertisedLocator.getPort(),
+                    targetLocator.getHostName(), targetLocator.getPort(), retryAttempt, exception});
+          }
+        }
+
+        return false;
+      }
+    }
+
+    @Override
+    public void run() {
+      Map<DistributionLocatorId, Set<LocatorJoinMessage>> failedMessages = new HashMap<>();
+      for (Map.Entry<Integer, Set<DistributionLocatorId>> entry : remoteLocators.entrySet()) {
+        for (DistributionLocatorId value : entry.getValue()) {
+          // Notify known remote locator about the advertised locator.
+          LocatorJoinMessage advertiseNewLocatorMessage = new LocatorJoinMessage(
+              joiningLocatorDistributedSystemId, joiningLocator, localLocatorId, "");
+          sendMessage(value, advertiseNewLocatorMessage, failedMessages);
+
+          // Notify the advertised locator about remote known locator.
+          LocatorJoinMessage advertiseKnownLocatorMessage =
+              new LocatorJoinMessage(entry.getKey(), value, localLocatorId, "");
+          sendMessage(joiningLocator, advertiseKnownLocatorMessage, failedMessages);
+        }
+      }
+
+      // Retry failed messages and remove those that succeed.
+      if (!failedMessages.isEmpty()) {
+        for (int attempt = 1; attempt <= WAN_LOCATOR_DISTRIBUTION_RETRY_ATTEMPTS; attempt++) {
+
+          for (Map.Entry<DistributionLocatorId, Set<LocatorJoinMessage>> entry : failedMessages
+              .entrySet()) {
+            DistributionLocatorId targetLocator = entry.getKey();
+            Set<LocatorJoinMessage> joinMessages = entry.getValue();
+
+            for (LocatorJoinMessage locatorJoinMessage : joinMessages) {
+              if (retryMessage(targetLocator, locatorJoinMessage, attempt)) {
```

Comment:
Do we need to check; if this locator is left distributed system during retry?

Reply:
We don't expect a reply to the message and the target locator might not even be part of the distributed system from which the message is sent so no, I don't think we need to check if it left the distributed system. Please let me know if you think otherwise.